Downstream encryption:fix-encrypted-version for repairing "bad signature" errors

PVince81 opened the pull request:
For fixing "Bad signature" errors.

Imported from https://github.com/owncloud/encryption/blob/c8d67f5/lib/Command/FixEncryptedVersion.php (master from that time).
No changes were made in the initial import.

### Tests

To test, upload a file and then manually change its "encrypted" value to either value - 1 or value + 1 then run `occ encryption:fix-encrypted-version`

#### User-key encryption

- ~~recovering all files for user~~ not supported
- ~~recovering by path for user~~ not supported

#### Master-key encryption

- [x] recovering all files for user
- [x] recovering by path for user

#### Other

- [x] does it repair when "encryption_skip_signature_check" is set ? :no_entry_sign: see below
- [x] group folders ??? => not supported

PVince81 committed:
Downstream encryption:fix-encrypted-version

For fixing "Bad signature" errors.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

juliushaertl commented:
>  group folders

Groupfolders currently don't support encryption https://github.com/nextcloud/groupfolders/commit/4521451cbb68b8b0760cd7603f6a3fc2482353f7

PVince81 commented:
- [ ] BUG: does not fix the files when signature check is enabled, because the verification thinks the files are ok
   - solution 1: detect the option and ask the admin to disable it, or disable it temporarily.
      - pros: easy to implement, non-intrusive
      - cons: if admin disables it it might make the instance unusable for users for which bad signature errors existed until the repair happens => but maybe here we should advice to only run this command in single-user mode ??
   - solution 2: disable the option while the command runs
      - cons: dangerous as it will affect concurrent requests, like in solution 1
      - cons: messes up with config.php, might not work on clustered setups
   - solution 3: adjust the "skip signature" check code to also allow for another flag to be passed it to disable skipping only within the occ command run
      - pros: safest approach
      - cons: intrusive as it requires more modifications (could potentially make backporting more difficult)

Thoughts ?

PVince81 commented:
rebased as I forgot to pull before I made this branch

PVince81 committed:
Downstream FixEncryptedVersionTest

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I've now also downstreamed the matching unit tests file.

PVince81 committed:
Detect disabled signature check when reparing

When running occ encryption:fix-encrypted-version, detect whether the
setting 'encryption_skip_signature_check' is set and abort if it is,
because the repair cannot detect version mismatch errors with it
enabled.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
for https://github.com/nextcloud/server/pull/27638#issuecomment-866920679 I went with approach 1: display an error if the skip flag is set. That should do for now.

PVince81 committed:
Fix warnings in FixEncryptedVersion command

Fixed code warnings

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Fix FixEncryptedVersionTest test

Fixed setup to use EncryptionTrait like other existing tests.
Fix expectations to not rely on side effects from previous test cases.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
I'm trying to fix the unit tests, there seem to be a discrepancy of behavior, but could be due to test setup

PVince81 commented:
okay, I'm guessing that NC doesn't encrypt / doesn't increase the encryption version when a file is empty, so the test calls to `touch()` don't increase the version

PVince81 commented:
yeah, I suspected that:
```

OCA\Encryption\Tests\Command\FixEncryptedVersionTest::testEncryptedVersionLessThanOriginalValue
--


OCA\Encryption\Exceptions\MultiKeyDecryptException:
 multikeydecrypt with share key failed:error:0E06D06C:configuration file
 routines:NCONF_get_string:no value
```

the test setup is dirty and causing side effects

- [ ] redo test setup for encryption to avoid side effects with other tests (use more mocks?)
- [ ] php:cs fix

PVince81 commented:
the tests of the app run fine for me locally, so I suspect that there might be another test that runs before that doesn't clean up properly and is causing side effects. good luck finding which one :cry:

PVince81 commented:
I've done some manual bisecting by deleting tests and running the whole DB suite locally: when deleting "apps/dav/tests" the encryption tests pass. So that folder likely contains a test that is causing side effects.
Digging deeper... :hole:

PVince81 commented:
getting closer, it's in the "apps/dav/tests/unit/Connector/Sabre/RequestTest" folder. If I delete that folder the encryption test will pass.

aha, I see some encryption-related tests in there, so probably the setup there isn't cleaning up fully

PVince81 commented:
So it seems the problem is not directly a side effect, but the way the setup is done. In the latter class, the `EncryptionTrait` is used with some minor gymnastics.

I've now adjusted the test class to do the same gymnastics and the side effect is gone.

PVince81 commented:
all green now, please review @juliushaertl @icewind1991

juliusknorr reviewed (approved):
Tested and works and code looks good.

juliushaertl commented:
Since the user key todos are still in the description, I'd say user key handling is out of scope for now since we'd need to provide the password of the user somehow. If we might need a recover option for that, we could probably maybe wrap the fixing logic into a small app to allow users to trigger it within their session as a kind of self service.

PVince81 commented:
> Since the user key todos are still in the description, I'd say user key handling is out of scope for now since we'd need to provide the password of the user somehow. If we might need a recover option for that, we could probably maybe wrap the fixing logic into a small app to allow users to trigger it within their session as a kind of self service.

Indeed. IIRC user key decryption is not supported by the command anyway as it would require either the user's password or a recovery key to be configured for that user.

juliushaertl commented:
I didn't see any check in that regard so i guess it would still try to find the matching encryptedVersion which would just not be successful then, but also haven't tested that.

PVince81 commented:
> I didn't see any check in that regard so i guess it would still try to find the matching encryptedVersion which would just not be successful then, but also haven't tested that.

if this is a concern, it should be possible to add a check for master key for this command and abort in case none is found

all this is assuming that the "default encryption module" is being used

PVince81 committed:
Prevent running FixEncryptedVersion without master key

Return an error when running occ encryption:fix-encrypted-version
when master key encryption is not enabled.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
@juliushaertl I've added an extra check https://github.com/nextcloud/server/pull/27638/commits/d3eeecba54fc891da20093c88648e5f10ed0e706

please re-review

PVince81 commented:
failing test unrelated

PVince81 commented:
/backport to stable22

PVince81 commented:
While this is an addition / feature, I wonder if we should backport this further down as it will be useful for fixing broken setups and would be safer than letting people patch their instances there.

juliushaertl commented:
> While this is an addition / feature, I wonder if we should backport this further down as it will be useful for fixing broken setups and would be safer than letting people patch their instances there.

I'd vote for that :+1: Any concers @skjnldsv ?

PVince81 commented:
force-merge needed due to intermitted acceptance test failure

skjnldsv commented:
/backport to stable21

skjnldsv commented:
/backport to stable20

skjnldsv commented:
/backport to stable22

PVince81 commented:
PR for the docs here: https://github.com/nextcloud/documentation/pull/6884
